File: Integer/RecoverRotateSortedArray/RecoveryRatateSortedArray.cc
class RecoveryRotateSortedArray {
	public:
		void recoveryRotateSortedArray(vector<int> &nums) {
			
			for(vector<int>::size_type i = 0; i < nums.size(); i++) {
				if(nums[i] > nums[i + 1]) {
					rotateHelper(nums, 0, i);
					rotateHelper(nums, i + 1, nums.size() - 1);
					rotateHelper(nums, 0, nums.size() - 1);
					return;
				}
			}
	}
	
	private:
		void rotateHelper(vector<int> &nums, vector<int>::size_type start, vector<int>::size_type end) {
			while(vector<int>::size_type i = start ,j = end; i < j; ++i, --j) {
				int tmp = nums[i];
				nums[i] = nums[j];
				nums[j] = tmp;
			}
	}
}

File: Integer/RecoverRotateSortedArray/RecoveryRatateSortedArray.java
class RecoveryRotateSortedArray {
	public void recoveryRotateSortedArray(ArrayList<Integer> nums) {
		if(nums == null || nums.size() == 0)
			return;
		
		int pos = 1;
		while(pos < nums.size()) {
			if(nums.get(pos - 1) > nums.get(pos))
				break;
			pos++;
		}
		
		rotateHelper(nums, 0, pos - 1);
		rotateHelper(nums, pos, nums.size() - 1);
		rotateHelper(nums, 0, nums.size() - 1);
	}
	
	private void rotateHelper(ArrayList<Integer> nums, int left, int right) {
		while(left < right) {
			int tmp = nums.get(left);
			nums.set(left, nums.get(right));
			nums.set(right, tmp);
			left++;
			right--;
		}
	}
}